File: python/day15.py
import heapq

def get_risk(board, x, y):
    w, h = len(board[0]), len(board)
    res = board[y % h][x % w]
    res += x // w + y // h
    if res > 9:
        res = res % 10 + 1
    return res

def get_min_risk(board, pos, goal):
    w, h = len(board[0]), len(board)
    to_explore = [(0, pos)]
    scores = {pos: 0}

    while len(to_explore) > 0:
        score, pos = heapq.heappop(to_explore)
        x, y = pos
        if x == goal[0] and y == goal[1]:
            return scores[pos]

        for dx, dy in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
            cx, cy = x + dx, y + dy
            if cx < 0 or cy < 0 or cx > goal[0] or cy > goal[1]:
                continue
            cpos = (cx, cy)
            cscore = score + get_risk(board, cx, cy)
            if cpos not in scores or scores[cpos] > cscore:
                scores[cpos] = cscore
                heapq.heappush(to_explore, (cscore, cpos))

def solution():
    board = [list(map(int, list(s))) for s in open("../data/15.txt").read().split()]
    w, h = len(board[0]), len(board)
    res1 = get_min_risk(board, (0, 0), (w - 1, h - 1))
    res2 = get_min_risk(board, (0, 0), (w * 5 - 1, h * 5 - 1))
    print(f"Answer 1: {res1}\nAnswer 2: {res2}")

File: rust/src/day15.rs
use crate::common;
use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::collections::HashMap;

#[derive(Copy, Clone, Eq, PartialEq)]
struct Node {
    score: usize,
    pos: (i32, i32),
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        other.score.cmp(&self.score)
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

const OFFS: [(i32, i32); 4] = [(0, 1), (0, -1), (1, 0), (-1, 0)];

fn get_risk(board: &Vec<Vec<u8>>, x: i32, y: i32) -> usize {
    let (w, h) = (board[0].len() as i32, board.len() as i32);
    let mut res = board[(y % h) as usize][(x % w) as usize] as usize;
    res += (x / w + y / h) as usize;
    if res > 9 {
        res = res % 10 + 1;
    }
    res
}

fn get_min_risk(board: &Vec<Vec<u8>>, pos: (i32, i32), goal: (i32, i32)) -> usize {
    let mut to_explore = BinaryHeap::new();
    to_explore.push(Node { score: 0, pos: pos });

    let mut scores = HashMap::new();
    scores.insert(pos, 0usize);

    while let Some(Node { score, pos }) = to_explore.pop() {
        let (x, y) = pos;
        if pos == goal {
            return scores[&pos];
        }
        for (dx, dy) in OFFS {
            let (cx, cy) = (x + dx, y + dy);
            if cx < 0 || cy < 0 || cx > goal.0 || cy > goal.1 {
                continue;
            }
            let cpos = (cx, cy);
            let cscore = score + get_risk(&board, cx, cy);
            if !scores.contains_key(&cpos) || scores[&cpos] > cscore {
                scores.insert(cpos, cscore);
                to_explore.push(Node {
                    score: cscore,
                    pos: cpos,
                });
            }
        }
    }
    0
}

pub(crate) fn solution() {
    let lines = common::read_lines(&common::data_file(15)).unwrap();
    let board: Vec<Vec<u8>> = lines
        .iter()
        .map(|x| x.as_bytes().iter().map(|c| c - '0' as u8).collect())
        .collect();
    let (w, h) = (board[0].len() as i32, board.len() as i32);
    let res1 = get_min_risk(&board, (0, 0), (w - 1, h - 1));
    let res2 = get_min_risk(&board, (0, 0), (w * 5 - 1, h * 5 - 1));
    println!("Answer 1: {:?}\nAnswer 2: {:?}", res1, res2);
}
